ocaml/libs/tracing/tracing.ml

```diff
@@ -0,0 +1,619 @@
+    | _ ->
+        span
+
+  let to_string s = Rpcmarshal.marshal t.Rpc.Types.ty s |> Jsonrpc.to_string
+
+  let of_string s =
+    Jsonrpc.of_string s
+    |> Rpcmarshal.unmarshal t.Rpc.Types.ty
+    |> Result.to_option
+end
+
+module Spans = struct
+  let lock = Mutex.create ()
+
+  let spans = Hashtbl.create 100
+
+  let max_spans = ref 1000
+
+  let set_max_spans x = max_spans := x
+
+  let max_traces = ref 1000
+
+  let set_max_traces x = max_traces := x
+
+  let finished_spans = Hashtbl.create 100
+
+  let span_hashtbl_is_empty () = Hashtbl.length spans = 0
+
+  let add_to_spans ~(span : Span.t) =
+    Xapi_stdext_threads.Threadext.Mutex.execute lock (fun () ->
+        let key = span.context.trace_id in
+        match Hashtbl.find_opt spans key with
+        | None ->
+            if Hashtbl.length spans < !max_traces then
+              Hashtbl.add spans key [span]
+        | Some span_list ->
+            if List.length span_list < !max_spans then
+              Hashtbl.replace spans key (span :: span_list)
+    )
+
+  let mark_finished ~(span : Span.t) =
+    Xapi_stdext_threads.Threadext.Mutex.execute lock (fun () ->
+        let key = span.context.trace_id in
+        match Hashtbl.find_opt spans key with
+        | None ->
+            debug "Span does not exist or already finished"
+        | Some span_list -> (
+            ( match
+                List.filter (fun x -> x.Span.context <> span.context) span_list
+              with
+            | [] ->
+                Hashtbl.remove spans key
+            | filtered_list ->
+                Hashtbl.replace spans key filtered_list
+            ) ;
+            match Hashtbl.find_opt finished_spans key with
+            | None ->
+                if Hashtbl.length finished_spans < !max_traces then
+                  Hashtbl.add finished_spans key [span]
+            | Some span_list ->
+                if List.length span_list < !max_spans then
+                  Hashtbl.replace finished_spans key (span :: span_list)
+          )
+    )
+
+  let span_is_finished x =
+    match x with
+    | None ->
+        false
+    | Some (span : Span.t) -> (
+      match Hashtbl.find_opt finished_spans span.context.trace_id with
+      | None ->
+          false
+      | Some span_list ->
+          List.exists (fun x -> x = span) span_list
+    )
+
+  (** since copies the existing finished spans and then clears the existing spans as to only export them once  *)
+  let since () =
+    Xapi_stdext_threads.Threadext.Mutex.execute lock (fun () ->
+        let copy = Hashtbl.copy finished_spans in
+        Hashtbl.clear finished_spans ;
+        copy
+    )
+
+  let dump () = Hashtbl.(copy spans, Hashtbl.copy finished_spans)
+
+  module GC = struct
+    let lock = Mutex.create ()
+
+    let span_timeout = ref 86400.
+
+    let span_timeout_thread = ref None
+
+    let gc_inactive_spans () =
+      Xapi_stdext_threads.Threadext.Mutex.execute lock (fun () ->
+          Seq.iter
```

Comment:
This still has the same problem that the hashtable is being modified as you're iterating through it. Seq.iter is sort of a lazy iterator fetching elements one at a time.

Reply:
What would be a good way to get around this issue

Reply:
Try `Hashbtl.filter_map_inplace`: it safely iterates over all elements and you get to decide whether to keep it or not by returning 'Some v' or 'None'.
Also check the logic, because it looks like only the debug message is under the 'if' scope and mark_as_finished is always called on all elements currently, was that intended?

Reply:
I have added that in. I divided `marked_finished` into two functions, one of which removes the span from the hash-table and one that adds it to the second hash-table. 4d0cbfa2b108943df5691958342af6f7b7d42511 for the new implementation.